src/topp/SageAdapter.cpp

```diff
@@ -0,0 +1,547 @@
+   {
+     origin += res->getOneLetterCode();
+   }
+
+    return String("\"") + origin + "\": " + String(mod->getDiffMonoMass());
+  }
+
+  // formats all mod entries into a single multi-line json string
+  String getModDetailsString(const OpenMS::ModifiedPeptideGenerator::MapToResidueType& mod_map)
+  {
+    String mod_details;
+    for (auto it = mod_map.val.begin(); it != mod_map.val.end(); ++it)
+    {
+      const auto& mod = it->first;
+      const auto& res = it->second;
+      mod_details += getModDetails(mod, res);     
+      if (std::next(it) != mod_map.val.end())
+      {
+        mod_details += ",\n";
+      }
+    }
+    return mod_details;
+  }
+
+  // impute values into config_template
+  String imputeConfigIntoTemplate()
+  {
+    String config_file = config_template;
+    config_file.substitute("##bucket_size##", String(getIntOption_("bucket_size")));
+    config_file.substitute("##min_len##", String(getIntOption_("min_len")));
+    config_file.substitute("##max_len##", String(getIntOption_("max_len")));
+    config_file.substitute("##missed_cleavages##", String(getIntOption_("missed_cleavages")));
+    config_file.substitute("##fragment_min_mz##", String(getDoubleOption_("fragment_min_mz")));
+    config_file.substitute("##fragment_max_mz##", String(getDoubleOption_("fragment_max_mz")));
+    config_file.substitute("##peptide_min_mass##", String(getDoubleOption_("peptide_min_mass")));
+    config_file.substitute("##peptide_max_mass##", String(getDoubleOption_("peptide_max_mass")));
+    config_file.substitute("##min_ion_index##", String(getIntOption_("min_ion_index")));
+    config_file.substitute("##max_variable_mods##", String(getIntOption_("max_variable_mods")));
+    config_file.substitute("##precursor_tol_unit##", precursor_tol_unit  == "Da" ? "da" : "ppm"); // sage might expect lower-case "da"
+    config_file.substitute("##precursor_tol_left##", String(getDoubleOption_("precursor_tol_left")));
+    config_file.substitute("##precursor_tol_right##", String(getDoubleOption_("precursor_tol_right")));
+    config_file.substitute("##fragment_tol_unit##", fragment_tol_unit == "Da" ? "da" : "ppm"); // sage might expect lower-case "da"
+    config_file.substitute("##fragment_tol_left##", String(getDoubleOption_("fragment_tol_left")));
+    config_file.substitute("##fragment_tol_right##", String(getDoubleOption_("fragment_tol_right")));
+    String isotope_errors = String(getIntList_("isotope_errors")).remove('[').remove(']');
+    config_file.substitute("##isotope_errors##", isotope_errors);
+    config_file.substitute("##min_matched_peaks##", String(getIntOption_("min_matched_peaks")));
+    config_file.substitute("##report_psms##", String(getIntOption_("report_psms")));
+    config_file.substitute("##decoy_tag##", String(getStringOption_("decoy_prefix")));
+
+    String enzyme = getStringOption_("enzyme");
+    String enzyme_details;
+    if (enzyme == "Trypsin")
```

Comment:
How about adding those as functions to Enzyme.h?

Reply:
Hmm Sage has no real notion of an enzyme name or id. Just basically the cutting rules etc. how would you add that?

Reply:
We also have the cutting rules in our enzyme db. Just return them in a way that Sage accepts.

Reply:
Hmm makes things a bit more convoluted without any notable gain.
E.g., one needs to encode the sage input as e.g., comma separated string, parse that one and impute the values in the json.
There is no sage output that contains the enzyme with a distinct ID, so we can't really make use of putting that into EnzymeDB (e.g., for ID file conversion). In that case I would rather keep it out of EnzymesDB as this adds some dependency to a third-party tool that is not really needed.

Reply:
I mean:
enz.getCleavageAAs()
enz.getRestrictedAAs()
enz.isCTerminal()

I am pretty sure this kind of logic is needlessly replicated somewhere else already. See PercolatorAdapter or PSMFeatureExtractor.
I don't think this is third-party dependent at all.

Reply:
Our enzyme class only uses RegEx which is hard to convert. I think you mean code like: https://github.com/OpenMS/OpenMS/blob/e829f8bbcba9ea070b71f33a5d508a51b12c4003/src/topp/PercolatorAdapter.cpp#L319-L382 this doesn't really help here.

Reply:
these functions don't exist. we read the regex from the xml

Reply:
That's why I think we should add them, so we can use them in both places instead of basically replicating the whole enzyme database.